File: examples-salvo/semi-complex/moonix/05_kmp.lua
local text, pat = "ababcabcabababd", "ababd"
local lps = { 0 }
local j = 0
for i = 2, #pat do
	while j > 0 and pat:sub(i, i) ~= pat:sub(j + 1, j + 1) do
		j = lps[j]
	end
	if pat:sub(i, i) == pat:sub(j + 1, j + 1) then
		j = j + 1
	end
	lps[i] = j
end
local i = 1
j = 0
local found = -1
while i <= #text do
	if text:sub(i, i) == pat:sub(j + 1, j + 1) then
		i = i + 1
		j = j + 1
		if j == #pat then
			found = i - #pat
			break
		end
	elseif j > 0 then
		j = lps[j]
	else
		i = i + 1
	end
end
print(found)

File: examples-salvo/semi-complex/cephyr/05_kmp.c
#include <stdio.h>
#include <string.h>
int
main (void)
{
  const char *t = "ababcabcabababd", *p = "ababd";
  int l[5] = { 0 }, j = 0, i;
  for (i = 1; i < 5; i++)
    {
      while (j && p[i] != p[j])
        j = l[j - 1];
      if (p[i] == p[j])
        j++;
      l[i] = j;
    }
  j = 0;
  for (i = 0; t[i]; i++)
    {
      while (j && t[i] != p[j])
        j = l[j - 1];
      if (t[i] == p[j])
        j++;
      if (j == 5)
        {
          printf ("%d\n", i - 4);
          return 0;
        }
    }
  printf ("-1\n");
}
